File: app/src/androidTest/java/org/tasks/data/TagDataDaoTest.java
package org.tasks.data;

import static com.google.common.collect.Sets.newHashSet;
import static com.natpryce.makeiteasy.MakeItEasy.with;
import static java.util.Collections.singletonList;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.tasks.makers.TagDataMaker.NAME;
import static org.tasks.makers.TagDataMaker.UID;
import static org.tasks.makers.TagDataMaker.newTagData;
import static org.tasks.makers.TagMaker.TAGDATA;
import static org.tasks.makers.TagMaker.TAGUID;
import static org.tasks.makers.TagMaker.TASK;
import static org.tasks.makers.TaskMaker.ID;
import static org.tasks.makers.TaskMaker.newTask;

import androidx.core.util.Pair;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import com.google.common.collect.ImmutableList;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import java.util.Set;
import javax.inject.Inject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.tasks.injection.InjectingTestCase;
import org.tasks.injection.TestComponent;
import org.tasks.makers.TagMaker;

@RunWith(AndroidJUnit4.class)
public class TagDataDaoTest extends InjectingTestCase {

  @Inject TaskDao taskDao;
  @Inject TagDao tagDao;
  @Inject TagDataDao tagDataDao;

  @Test
  public void tagDataOrderedByNameIgnoresNullNames() {
    tagDataDao.createNew(newTagData(with(NAME, (String) null)));

    assertTrue(tagDataDao.tagDataOrderedByName().isEmpty());
  }

  @Test
  public void tagDataOrderedByNameIgnoresEmptyNames() {
    tagDataDao.createNew(newTagData(with(NAME, "")));

    assertTrue(tagDataDao.tagDataOrderedByName().isEmpty());
  }

  @Test
  public void getTagWithCaseForMissingTag() {
    assertEquals("derp", tagDataDao.getTagWithCase("derp"));
  }

  @Test
  public void getTagWithCaseFixesCase() {
    tagDataDao.createNew(newTagData(with(NAME, "Derp")));

    assertEquals("Derp", tagDataDao.getTagWithCase("derp"));
  }

  @Test
  public void getTagsByName() {
    TagData tagData = newTagData(with(NAME, "Derp"));
    tagDataDao.createNew(tagData);

    assertEquals(singletonList(tagData), tagDataDao.getTags(singletonList("Derp")));
  }

  @Test
  public void getTagsByNameCaseSensitive() {
    tagDataDao.createNew(newTagData(with(NAME, "Derp")));

    assertTrue(tagDataDao.getTags(singletonList("derp")).isEmpty());
  }

  @Test
  public void getTagDataForTask() {
    Task taskOne = newTask();
    Task taskTwo = newTask();
    taskDao.createNew(taskOne);
    taskDao.createNew(taskTwo);

    TagData tagOne = newTagData(with(NAME, "one"));
    TagData tagTwo = newTagData(with(NAME, "two"));
    tagDataDao.createNew(tagOne);
    tagDataDao.createNew(tagTwo);

    tagDao.insert(TagMaker.newTag(with(TAGDATA, tagOne), with(TASK, taskOne)));
    tagDao.insert(TagMaker.newTag(with(TAGDATA, tagTwo), with(TASK, taskTwo)));

    assertEquals(singletonList(tagOne), tagDataDao.getTagDataForTask(taskOne.getId()));
  }

  @Test
  public void getEmptyTagSelections() {
    Pair<Set<String>, Set<String>> selections = tagDataDao.getTagSelections(ImmutableList.of(1L));
    assertTrue(selections.first.isEmpty());
    assertTrue(selections.second.isEmpty());
  }

  @Test
  public void getPartialTagSelections() {
    newTag(1, "tag1", "tag2");
    newTag(2, "tag2", "tag3");


    assertEquals(
        newHashSet("tag1", "tag3"), tagDataDao.getTagSelections(ImmutableList.of(1L, 2L)).first);
  }

  @Test
  public void getEmptyPartialSelections() {
    newTag(1, "tag1");
    newTag(2, "tag1");

    assertTrue(tagDataDao.getTagSelections(ImmutableList.of(1L, 2L)).first.isEmpty());
  }

  @Test
  public void getCommonTagSelections() {
    newTag(1, "tag1", "tag2");
    newTag(2, "tag2", "tag3");

    assertEquals(
        newHashSet("tag2"), tagDataDao.getTagSelections(ImmutableList.of(1L, 2L)).second);
  }

  @Test
  public void getEmptyCommonSelections() {
    newTag(1, "tag1");
    newTag(2, "tag2");

    assertTrue(tagDataDao.getTagSelections(ImmutableList.of(1L, 2L)).second.isEmpty());
  }

  private void newTag(long taskId, String... tags) {
    Task task = newTask(with(ID, taskId));
    taskDao.createNew(task);
    for (String tag : tags) {
      tagDao.insert(TagMaker.newTag(with(TASK, task), with(TAGUID, tag)));
    }
  }

  @Override
  protected void inject(TestComponent component) {
    component.inject(this);
  }
}

File: app/src/androidTest/java/org/tasks/data/TagDataDaoTest.kt
package org.tasks.data

import com.natpryce.makeiteasy.MakeItEasy.with
import com.todoroo.astrid.dao.TaskDao
import dagger.hilt.android.testing.HiltAndroidTest
import dagger.hilt.android.testing.UninstallModules
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.tasks.data.dao.TagDao
import org.tasks.data.dao.TagDataDao
import org.tasks.data.entity.Tag
import org.tasks.data.entity.TagData
import org.tasks.injection.InjectingTestCase
import org.tasks.injection.ProductionModule
import org.tasks.makers.TaskMaker.ID
import org.tasks.makers.TaskMaker.newTask
import javax.inject.Inject

@UninstallModules(ProductionModule::class)
@HiltAndroidTest
class TagDataDaoTest : InjectingTestCase() {
    @Inject lateinit var taskDao: TaskDao
    @Inject lateinit var tagDao: TagDao
    @Inject lateinit var tagDataDao: TagDataDao

    @Test
    fun tagDataOrderedByNameIgnoresNullNames() = runBlocking {
        tagDataDao.insert(TagData(name = null))
        assertTrue(tagDataDao.tagDataOrderedByName().isEmpty())
    }

    @Test
    fun tagDataOrderedByNameIgnoresEmptyNames() = runBlocking {
        tagDataDao.insert(TagData(name = ""))
        assertTrue(tagDataDao.tagDataOrderedByName().isEmpty())
    }

    @Test
    fun getTagWithCaseForMissingTag() = runBlocking {
        assertEquals("derp", tagDataDao.getTagWithCase("derp"))
    }

    @Test
    fun getTagWithCaseFixesCase() = runBlocking {
        tagDataDao.insert(TagData(name = "Derp"))
        assertEquals("Derp", tagDataDao.getTagWithCase("derp"))
    }

    @Test
    fun getTagsByName() = runBlocking {
        val tagData = TagData(name = "Derp").let { it.copy(id = tagDataDao.insert(it)) }
        assertEquals(listOf(tagData), tagDataDao.getTags(listOf("Derp")))
    }

    @Test
    fun getTagsByNameCaseSensitive() = runBlocking {
        tagDataDao.insert(TagData(name = "Derp"))
        assertTrue(tagDataDao.getTags(listOf("derp")).isEmpty())
    }

    @Test
    fun getTagDataForTask() = runBlocking {
        val taskOne = newTask()
        val taskTwo = newTask()
        taskDao.createNew(taskOne)
        taskDao.createNew(taskTwo)
        val tagOne = TagData(name = "one").let { it.copy(id = tagDataDao.insert(it)) }
        val tagTwo = TagData(name = "two").let { it.copy(id = tagDataDao.insert(it)) }
        tagDao.insert(Tag(task = taskOne.id, taskUid = taskOne.uuid, tagUid = tagOne.remoteId))
        tagDao.insert(Tag(task = taskTwo.id, taskUid = taskTwo.uuid, tagUid = tagTwo.remoteId))
        assertEquals(listOf(tagOne), tagDataDao.getTagDataForTask(taskOne.id))
    }

    @Test
    fun getEmptyTagSelections() = runBlocking {
        val selections = tagDataDao.getTagSelections(listOf(1L))
        assertTrue(selections.first.isEmpty())
        assertTrue(selections.second.isEmpty())
    }

    @Test
    fun getPartialTagSelections() = runBlocking {
        newTag(1, "tag1", "tag2")
        newTag(2, "tag2", "tag3")
        assertEquals(
                setOf("tag1", "tag3"), tagDataDao.getTagSelections(listOf(1L, 2L)).first)
    }

    @Test
    fun getEmptyPartialSelections() = runBlocking {
        newTag(1, "tag1")
        newTag(2, "tag1")
        assertTrue(tagDataDao.getTagSelections(listOf(1L, 2L)).first.isEmpty())
    }

    @Test
    fun getCommonTagSelections() = runBlocking {
        newTag(1, "tag1", "tag2")
        newTag(2, "tag2", "tag3")
        assertEquals(setOf("tag2"), tagDataDao.getTagSelections(listOf(1L, 2L)).second)
    }

    @Test
    fun getEmptyCommonSelections() = runBlocking {
        newTag(1, "tag1")
        newTag(2, "tag2")
        assertTrue(tagDataDao.getTagSelections(listOf(1L, 2L)).second.isEmpty())
    }

    @Test
    fun getSelectionsWithNoTags() = runBlocking {
        newTag(1)
        val selections = tagDataDao.getTagSelections(listOf(1L))
        assertTrue(selections.first.isEmpty())
        assertTrue(selections.second.isEmpty())
    }

    @Test
    fun noCommonSelectionsWhenOneTaskHasNoTags() = runBlocking {
        newTag(1, "tag1")
        newTag(2)
        val selections = tagDataDao.getTagSelections(listOf(1L, 2L))
        assertEquals(setOf("tag1"), selections.first)
        assertTrue(selections.second.isEmpty())
    }

    private suspend fun newTag(taskId: Long, vararg tags: String) {
        val task = newTask(with(ID, taskId))
        taskDao.createNew(task)
        for (tag in tags) {
            tagDao.insert(Tag(task = task.id, taskUid = task.uuid, tagUid = tag))
        }
    }
}